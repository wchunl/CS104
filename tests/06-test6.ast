ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_STRUCT "struct" (7.5.1)
|   |   TOK_IDENT "foo" (7.5.8) variable
|   TOK_STRUCT "struct" (7.6.1)
|   |   TOK_IDENT "bar" (7.6.8) variable
|   TOK_FUNCTION "" (7.8.1) function
|   |   TOK_TYPE_ID "" (7.8.1) typeid
|   |   |   TOK_INT "int" (7.8.1) int
|   |   |   TOK_IDENT "f0" (7.8.5) variable
|   |   TOK_PARAM "(" (7.8.7) param
|   TOK_FUNCTION "" (7.9.1) function
|   |   TOK_TYPE_ID "" (7.9.1) typeid
|   |   |   TOK_INT "int" (7.9.1) int
|   |   |   TOK_IDENT "f1" (7.9.5) variable
|   |   TOK_PARAM "(" (7.9.8) param
|   |   |   TOK_TYPE_ID "" (7.9.9) typeid
|   |   |   |   TOK_INT "int" (7.9.9) int
|   |   |   |   TOK_IDENT "a" (7.9.13) int variable (7.9.9)
|   TOK_FUNCTION "" (7.10.1) function
|   |   TOK_TYPE_ID "" (7.10.1) typeid
|   |   |   TOK_INT "int" (7.10.1) int
|   |   |   TOK_IDENT "f2" (7.10.5) variable
|   |   TOK_PARAM "(" (7.10.8) param
|   |   |   TOK_TYPE_ID "" (7.10.9) typeid
|   |   |   |   TOK_INT "int" (7.10.9) int
|   |   |   |   TOK_IDENT "a" (7.10.13) int variable (7.9.9)
|   |   |   TOK_TYPE_ID "" (7.10.16) typeid
|   |   |   |   TOK_INT "int" (7.10.16) int
|   |   |   |   TOK_IDENT "b" (7.10.20) int variable (7.10.16)
|   TOK_FUNCTION "" (7.11.1) function
|   |   TOK_TYPE_ID "" (7.11.1) typeid
|   |   |   TOK_INT "int" (7.11.1) int
|   |   |   TOK_IDENT "f3" (7.11.5) variable
|   |   TOK_PARAM "(" (7.11.8) param
|   |   |   TOK_TYPE_ID "" (7.11.9) typeid
|   |   |   |   TOK_STRING "string" (7.11.9) string
|   |   |   |   TOK_IDENT "a" (7.11.16) int variable (7.9.9)
|   |   |   TOK_TYPE_ID "" (7.11.19) typeid
|   |   |   |   TOK_STRING "string" (7.11.19) string
|   |   |   |   TOK_IDENT "b" (7.11.26) int variable (7.10.16)
|   |   |   TOK_TYPE_ID "" (7.11.29) typeid
|   |   |   |   TOK_STRING "string" (7.11.29) string
|   |   |   |   TOK_IDENT "c" (7.11.36) string variable (7.11.29)
|   TOK_FUNCTION "" (7.12.1) function
|   |   TOK_TYPE_ID "" (7.12.1) typeid
|   |   |   TOK_INT "int" (7.12.1) int
|   |   |   TOK_IDENT "f4" (7.12.5) variable
|   |   TOK_PARAM "(" (7.12.8) param
|   |   |   TOK_TYPE_ID "" (7.12.9) typeid
|   |   |   |   TOK_PTR "ptr" (7.12.9)
|   |   |   |   |   TOK_TYPE_ID "foo" (7.12.20) typeid
|   |   |   |   TOK_IDENT "a" (7.12.25) int variable (7.9.9)
|   |   |   TOK_TYPE_ID "" (7.12.28) typeid
|   |   |   |   TOK_PTR "ptr" (7.12.28)
|   |   |   |   |   TOK_TYPE_ID "bar" (7.12.39) typeid
|   |   |   |   TOK_IDENT "b" (7.12.44) int variable (7.10.16)
|   TOK_FUNCTION "" (7.14.1) function
|   |   TOK_TYPE_ID "" (7.14.1) typeid
|   |   |   TOK_INT "int" (7.14.1) int
|   |   |   TOK_IDENT "main" (7.14.5) variable
|   |   TOK_PARAM "(" (7.14.9) param
|   |   TOK_BLOCK "{" (7.14.12)
|   |   |   TOK_VARDECL "=" (7.15.13) variable lval
|   |   |   |   TOK_STRING "string" (7.15.4) string
|   |   |   |   TOK_IDENT "s" (7.15.11) string variable (7.15.13)
|   |   |   |   TOK_STRINGCON ""a"" (7.15.15) string const
|   |   |   TOK_VARDECL "=" (7.16.21) variable lval
|   |   |   |   TOK_ARRAY "array" (7.16.4) array
|   |   |   |   |   TOK_TYPE_ID "string" (7.16.10) typeid
|   |   |   |   TOK_IDENT "sa" (7.16.18) string variable (7.16.21)
|   |   |   |   TOK_ALLOC "alloc" (7.16.23)
|   |   |   |   |   TOK_ARRAY "array" (7.16.29) array
|   |   |   |   |   TOK_TYPE_ID "string" (7.16.35) typeid
|   |   |   |   |   TOK_INTCON "10" (7.16.44) int const
|   |   |   TOK_RETURN "return" (7.17.4)
|   |   |   |   TOK_INTCON "0" (7.17.11) int const
