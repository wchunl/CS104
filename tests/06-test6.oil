fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
          .struct foo
          .end
          .struct bar
          .end
f0:       .function int
          .end
f1:       .function int
          .param int a
          .end
f2:       .function int
          .param int a
          .param int b
          .end
f3:       .function int
          .param ptr a
          .param ptr b
          .param ptr c
          .end
f4:       .function int
          .param ptr a
          .param ptr b
          .end
main:     .function int
          .local ptr s
          .local ptr sa
          s = "a"
          malloc(4 * sizeof ptr)
          sa = $t0:i
          return 0
          return 
          .end
