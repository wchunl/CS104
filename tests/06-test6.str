string_set[   2]:   12597418120160561823 0x2078cf8->"int"
                    2594088158685378223 0x207d098->"10"
string_set[   4]:   15718330555005605877 0x207b268->"f2"
string_set[   7]:    9631199822919835226 0x207a878->"foo"
                   10959529184379665549 0x207ba98->"c"
string_set[   8]:   13497578857207247412 0x2078b98->"string"
string_set[  10]:    9201800968376702500 0x2079f88->"getchr"
string_set[  11]:    4993892634952068459 0x207b0b8->"a"
string_set[  14]:   15816506932648033945 0x2078ae8->"("
                    1396094175042499165 0x207c138->"<"
string_set[  15]:   12868445110721718657 0x2078ef8->")"
string_set[  17]:    6142509188972423790 0x2074738->""
string_set[  20]:    1583576014053361350 0x2079b58->"str"
string_set[  21]:    2336781128945484129 0x207ad38->"f0"
string_set[  26]:   12291392672316406109 0x20790b8->"line"
                    9611428142580219387 0x2074758->"fail"
                   11597697714117577063 0x207c958->"="
string_set[  28]:   18413295556186398326 0x20792c8->"putint"
string_set[  29]:   11911796782241914524 0x2074318->"void"
string_set[  33]:    6814028903552918164 0x207cab8->"array"
string_set[  35]:   11868120863286193964 0x207c318->"main"
string_set[  40]:   13945322250941355849 0x207a6f8->"struct"
string_set[  42]:   13686721134187162920 0x207aaa8->"}"
string_set[  44]:    6414370376350317282 0x207c8a8->"s"
string_set[  48]:    2451980591299959223 0x207bbf8->"f4"
string_set[  49]:   12679825168387489523 0x20799a8->"num"
string_set[  50]:      21606623943142546 0x2079268->"{"
string_set[  52]:    5344662657899890615 0x20790d8->";"
string_set[  54]:    7296690267450279874 0x2079588->"chr"
                   10838281452030117757 0x207b698->"b"
string_set[  55]:    2297668033614959926 0x207d168->"0"
                   17176659130743002837 0x207c088->"ptr"
string_set[  56]:   11474628671133349555 0x207ab78->"bar"
string_set[  63]:   14526079891305879776 0x207c2a8->">"
string_set[  64]:   16579916199511856337 0x207a488->"getln"
string_set[  65]:   10642371072705096219 0x207ca08->""a""
string_set[  66]:    6152274446545244381 0x207ae48->"f1"
string_set[  67]:    4706206694200543916 0x2078dc8->","
string_set[  73]:    9386170080689500517 0x2078c78->"expr"
string_set[  76]:   11166701219533395393 0x207a1f8->"getstr"
string_set[  80]:    8290819294854170237 0x207cd48->"alloc"
                    7211130878751977441 0x2079418->"putchr"
                    7115051347039519656 0x207cc98->"sa"
string_set[  83]:    5385639732984164590 0x207b538->"f3"
string_set[  85]:    4017852018407011111 0x2078ed8->"file"
string_set[  87]:   14635384904416252381 0x207cad8->"return"
string_set[  89]:   13348923631108115385 0x2079738->"putstr"
load_factor = 0.485
bucket_count = 97
max_bucket_size = 3
