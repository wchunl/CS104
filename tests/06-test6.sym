fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

foo (7.5.1) {0} struct foo

bar (7.6.1) {0} struct bar

f0 (7.8.1) {7} int function

f1 (7.9.1) {8} int function
   a (7.9.9) {1} int variable param lval 0

f2 (7.10.1) {9} int function
   a (7.10.9) {1} int variable param lval 0
   b (7.10.16) {1} int variable param lval 1

f3 (7.11.1) {10} int function
   a (7.11.9) {1} string variable param lval 0
   b (7.11.19) {1} string variable param lval 1
   c (7.11.29) {1} string variable param lval 2

f4 (7.12.1) {11} int function
   a (7.12.9) {1} ptr <struct foo> variable param lval 0
   b (7.12.28) {1} ptr <struct bar> variable param lval 1

main (7.14.1) {12} int function
   s (7.15.13) {1} string variable local lval 0
   sa (7.16.21) {1} array <string> variable local lval 1
