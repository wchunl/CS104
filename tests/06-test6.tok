#  1 "06-test6.oc"
#  1 "<built-in>"
#  1 "<command-line>"
#  1 "06-test6.oc"
#  1 "oclib.h"
# 22 "oclib.h"
   6 22.00001 261 TOK_VOID      void
   6 22.00006 282 TOK_IDENT     fail
   6 22.00011  40 '('           (
   6 22.00012 263 TOK_STRING    string
   6 22.00019 282 TOK_IDENT     expr
   6 22.00023  44 ','           ,
   6 22.00025 263 TOK_STRING    string
   6 22.00032 282 TOK_IDENT     file
   6 22.00036  44 ','           ,
   6 22.00038 262 TOK_INT       int
   6 22.00042 282 TOK_IDENT     line
   6 22.00046  41 ')'           )
   6 22.00047  59 ';'           ;
   6 24.00001 261 TOK_VOID      void
   6 24.00006 282 TOK_IDENT     putchr
   6 24.00013  40 '('           (
   6 24.00014 262 TOK_INT       int
   6 24.00018 282 TOK_IDENT     chr
   6 24.00021  41 ')'           )
   6 24.00022  59 ';'           ;
   6 25.00001 261 TOK_VOID      void
   6 25.00006 282 TOK_IDENT     putint
   6 25.00013  40 '('           (
   6 25.00014 262 TOK_INT       int
   6 25.00018 282 TOK_IDENT     num
   6 25.00021  41 ')'           )
   6 25.00022  59 ';'           ;
   6 26.00001 261 TOK_VOID      void
   6 26.00006 282 TOK_IDENT     putstr
   6 26.00013  40 '('           (
   6 26.00014 263 TOK_STRING    string
   6 26.00021 282 TOK_IDENT     str
   6 26.00024  41 ')'           )
   6 26.00025  59 ';'           ;
   6 28.00001 262 TOK_INT       int
   6 28.00005 282 TOK_IDENT     getchr
   6 28.00011  40 '('           (
   6 28.00012  41 ')'           )
   6 28.00013  59 ';'           ;
   6 29.00001 263 TOK_STRING    string
   6 29.00008 282 TOK_IDENT     getstr
   6 29.00014  40 '('           (
   6 29.00015  41 ')'           )
   6 29.00016  59 ';'           ;
   6 30.00001 263 TOK_STRING    string
   6 30.00008 282 TOK_IDENT     getln
   6 30.00013  40 '('           (
   6 30.00014  41 ')'           )
   6 30.00015  59 ';'           ;
#  4 "06-test6.oc"
   7 5.00001 269 TOK_STRUCT    struct
   7 5.00008 282 TOK_IDENT     foo
   7 5.00012 123 '{'           {
   7 5.00013 125 '}'           }
   7 5.00014  59 ';'           ;
   7 6.00001 269 TOK_STRUCT    struct
   7 6.00008 282 TOK_IDENT     bar
   7 6.00012 123 '{'           {
   7 6.00013 125 '}'           }
   7 6.00014  59 ';'           ;
   7 8.00001 262 TOK_INT       int
   7 8.00005 282 TOK_IDENT     f0
   7 8.00007  40 '('           (
   7 8.00008  41 ')'           )
   7 8.00009  59 ';'           ;
   7 9.00001 262 TOK_INT       int
   7 9.00005 282 TOK_IDENT     f1
   7 9.00008  40 '('           (
   7 9.00009 262 TOK_INT       int
   7 9.00013 282 TOK_IDENT     a
   7 9.00014  41 ')'           )
   7 9.00015  59 ';'           ;
   7 10.00001 262 TOK_INT       int
   7 10.00005 282 TOK_IDENT     f2
   7 10.00008  40 '('           (
   7 10.00009 262 TOK_INT       int
   7 10.00013 282 TOK_IDENT     a
   7 10.00014  44 ','           ,
   7 10.00016 262 TOK_INT       int
   7 10.00020 282 TOK_IDENT     b
   7 10.00021  41 ')'           )
   7 10.00022  59 ';'           ;
   7 11.00001 262 TOK_INT       int
   7 11.00005 282 TOK_IDENT     f3
   7 11.00008  40 '('           (
   7 11.00009 263 TOK_STRING    string
   7 11.00016 282 TOK_IDENT     a
   7 11.00017  44 ','           ,
   7 11.00019 263 TOK_STRING    string
   7 11.00026 282 TOK_IDENT     b
   7 11.00027  44 ','           ,
   7 11.00029 263 TOK_STRING    string
   7 11.00036 282 TOK_IDENT     c
   7 11.00037  41 ')'           )
   7 11.00038  59 ';'           ;
   7 12.00001 262 TOK_INT       int
   7 12.00005 282 TOK_IDENT     f4
   7 12.00008  40 '('           (
   7 12.00009 274 TOK_PTR       ptr
   7 12.00012  60 '<'           <
   7 12.00013 269 TOK_STRUCT    struct
   7 12.00020 282 TOK_IDENT     foo
   7 12.00023  62 '>'           >
   7 12.00025 282 TOK_IDENT     a
   7 12.00026  44 ','           ,
   7 12.00028 274 TOK_PTR       ptr
   7 12.00031  60 '<'           <
   7 12.00032 269 TOK_STRUCT    struct
   7 12.00039 282 TOK_IDENT     bar
   7 12.00042  62 '>'           >
   7 12.00044 282 TOK_IDENT     b
   7 12.00045  41 ')'           )
   7 12.00046  59 ';'           ;
   7 14.00001 262 TOK_INT       int
   7 14.00005 282 TOK_IDENT     main
   7 14.00009  40 '('           (
   7 14.00010  41 ')'           )
   7 14.00012 123 '{'           {
   7 15.00004 263 TOK_STRING    string
   7 15.00011 282 TOK_IDENT     s
   7 15.00013  61 '='           =
   7 15.00015 285 TOK_STRINGCON "a"
   7 15.00018  59 ';'           ;
   7 16.00004 271 TOK_ARRAY     array
   7 16.00009  60 '<'           <
   7 16.00010 263 TOK_STRING    string
   7 16.00016  62 '>'           >
   7 16.00018 282 TOK_IDENT     sa
   7 16.00021  61 '='           =
   7 16.00023 273 TOK_ALLOC     alloc
   7 16.00028  60 '<'           <
   7 16.00029 271 TOK_ARRAY     array
   7 16.00034  60 '<'           <
   7 16.00035 263 TOK_STRING    string
   7 16.00041  62 '>'           >
   7 16.00042  62 '>'           >
   7 16.00043  40 '('           (
   7 16.00044 283 TOK_INTCON    10
   7 16.00046  41 ')'           )
   7 16.00047  59 ';'           ;
   7 17.00004 268 TOK_RETURN    return
   7 17.00011 283 TOK_INTCON    0
   7 17.00012  59 ';'           ;
   7 18.00001 125 '}'           }
