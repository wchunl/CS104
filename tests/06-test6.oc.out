; 0.  1: # 1 "06-test6.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "06-test6.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "06-test6.oc" 2
; 7.  4: 
; 7.  5: struct foo {};
; 7.  6: struct bar {};
; 7.  7: 
; 7.  8: int f0();
; 7.  9: int f1 (int a);
; 7. 10: int f2 (int a, int b);
; 7. 11: int f3 (string a, string b, string c);
; 7. 12: int f4 (ptr<struct foo> a, ptr<struct bar> b);
; 7. 13: 
; 7. 14: int main() {
; 7. 15:    string s = "a";
; 7. 16:    array<string> sa = alloc<array<string>>(10);
; 7. 17:    return 0;
; 7. 18: }
